File: 1711. Count Good Meals/CountPairs.py
from typing import List
from collections import Counter


class Solution:
    
    def getPairs(self, val, count) -> int:
        
        answer = 0;
        for i in range(0, 22):
            power_of_two = 2**i
            target = power_of_two-val
            if target in count:
                answer += (count[target]-1) if target == val else count[target]
        
        return answer
        
    
    # O(n) time,
    # O(n) space,
    # Approach: hashtable, math
    def countPairs(self, deliciousness: List[int]) -> int:
        MOD = (10**9)+7
        answer = 0
        nums_count = Counter(deliciousness)
        for val in deliciousness:
            answer += self.getPairs(val, nums_count)
            
        answer //= 2
        answer %= MOD
        return answer

File: 1711. Count Good Meals/count_pairs.rs
use std::collections::HashMap;

impl Solution {
    
    fn getPairs(val: i32, count: &HashMap<i32, i32>) -> i64 {
        
        let mut answer = 0 as i64;
        
        for i in (0..22) {
            let power_of_two = i32::pow(2, i);
            let target = power_of_two - val;
            
            if count.contains_key(&target) {
                let target_count = *count.get(&target).unwrap() as i64;
                answer += if (target == val) {target_count-1} else {target_count};
            }   
        }
        return answer;
    }
    
    
    // O(n) time,
    // O(n) space,
    // Approach: hashmap, 
    pub fn count_pairs(deliciousness: Vec<i32>) -> i32 {
        let mut nums_count: HashMap<i32, i32> = HashMap::new();
        for num in &deliciousness {
            let mut count = 0;
            if nums_count.contains_key(&num) {
                count = *nums_count.get(&num).unwrap();
            }
            count += 1;
            nums_count.insert(*num, count);
        }
        
        let mut answer = 0;
        for num in deliciousness {
            answer += Solution::getPairs(num, &nums_count);
        }
        answer /= 2;
        let MOD = i64::pow(10, 9) + 7;
        answer %= MOD;
        return answer as i32;
    }
}